File: 백준/Gold/1662. 압축/압축.kt
class Decompressor(target: String) {
    private val s = target
    private var idx = 0

    fun decompress(): Int {
        var result = 0
        while (idx < s.length) {
            if (s[idx] == ')') {
                return result
            }
            result += if (s[idx] == '(') {
                (s[idx++ - 1].code - 48) * decompress() - 1
            } else {
                1
            }
            idx++
        }
        return result
    }
}

fun main(): Unit = with(System.`in`.bufferedReader()) {
    val dcp = Decompressor(readLine())
    println(dcp.decompress())
}

File: 백준/Gold/1662. 압축/압축.py
from sys import stdin


def decompress(s, idx):
    result = 0
    while idx < len(s):
        if s[idx] == ')':
            return result
        if s[idx] == '(':
            result += int(s[idx - 1]) * decompress(s, idx + 1) - 1
            count = 1
            while count > 0:
                idx += 1
                if s[idx] == '(':
                    count += 1
                elif s[idx] == ')':
                    count -= 1
        else:
            result += 1
        idx += 1
    return result


s = stdin.readline().strip()
print(decompress(s, 0))